### ISSUE TITLE
Using Close protocol message instead of DEALLOCATE to help prepared statements in PgBouncer

### ISSUE THREAD

**Issue by JelteF** [psycopg/psycopg#589]
PgBouncer does currently not parse queries at all, only protocol messages. But psycopg closes its statements using DEALOCATE sql call:

https://github.com/psycopg/psycopg/blob/63803637e5b20547807bcd632ce2487d871df44e/psycopg/psycopg/_preparing.py#L119

What do you think of about, instead of sending DEALLOCATE text query here, to instead send a Close protocol message? Because that would make it work automatically with the work-in-progress PgBouncer prepared statements patch:  https://github.com/pgbouncer/pgbouncer/pull/845

P.S. The DEALLOCATE ALL usage is similarly problematic, but maybe PgBouncer can make an exception for supporting that since it would be extremely easy to parse. And also it should not be necessary to use as I explained a few minutes ago on the PG mailing list thread that triggered this DEALLOCATE ALL usage (according to git blame): https://www.postgresql.org/message-id/flat/CAGECzQSUPTjfb7BYLg5c22tN4SSuLTRv9-6TpuZTpJ0q8hqp2w%40mail.gmail.com#b9b91ddf3d37160a3e8771e69b27f0ac

**Comment by dvarrazzo** [comment]
Prepared statements and pgbouncer are fundamentally incompatible. There is not expectation that they work. If you are using pgbouncer you must disable prepared statements.

We would to it ourselves, but PgBouncer offers no way to detect them: see https://github.com/pgbouncer/pgbouncer/issues/653

I also don't see how a Close message can replace DEALLOCATE.

**Comment by JelteF** [comment]
Colud you please reopen the issue? I probably wasn't clear on my friday end of day (and I updated my initial message a bit to hopefully clarify). But I'm a PgBouncer maintainers (although for less than a year). And I'm trying to suggest a (to me seemingly) small change in psycopg, so that it would work with the prepared statement support in transaction pooling mode that's currently actively being worked on for PgBouncer.

**Comment by JelteF** [comment]
> I also don't see how a Close message can replace DEALLOCATE.

Afaict `Close` is the protocol level equivalent of DEALLOCATE (except it does not support ALL). From the Postgres protocol docs (emphasis mine):
> The Close message closes an existing _prepared statement_ or portal and releases resources. 

Source: https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY

**Comment by dvarrazzo** [comment]
Ok, I thought Close closed the connection.

Does the libpq have a function to send a close? If not, then no, we cannot. We don't talk to the database at wire protocol level, only at libpq level.

As far as I see it, it is unwise to change psycopg to accommodate pgbouncer if you are developing an experimental feature. You are free to change psycopg code and propose a change but maybe only after your code is ready.

Also please know that I can see improvements possible in the communication between psycopg and pgbouncer, but if pgbouncer even refuses to identify itself and advertise its most important settings (does it support transactions or not?) then it's a non-starter.

**Comment by dvarrazzo** [comment]
For instance, just saying, we could even avoid to run DEALLOCATE at all on pgbouncer and rely that it will do its own garbage collection and dispose of connection after some use, so that prepared queries don't accumulate forever. But we could do that only if we see that we are connected to pgbouncer, and have some sort of "capabilities" object on the connection to manage the nuances.

Without https://github.com/pgbouncer/pgbouncer/issues/653 there's no place we can go. If the attitude of the project is to pretend that there is no difference between a pgbouncer connection vs. a native one, then you will have to deal with DEALLOCATE or whatever else we can legally throw at postgres, and assume unaltered semantics.

**Comment by JelteF** [comment]
After trying to quickly implement a POC of this myself. I realized it wasn't as easy as I thought because libpq doesn't actually expose an interface to send a `Close` message. So I submitted a patch for the next postgres commit fest to do so: https://www.postgresql.org/message-id/flat/CAMsr%2BYEsk0z6rctRiw54CkAehosPternSdrws%2Bf8LKMd4b4MHA%40mail.gmail.com#c29d117e648ef4fdf621f87a664dfd74

https://commitfest.postgresql.org/43/4373/

**Comment by dvarrazzo** [comment]
If it's merged, no problem using it in place of DEALLOCATE.

**Comment by JelteF** [comment]
It's merged now (for PG17): https://github.com/postgres/postgres/commit/28b5726561841556dc3e00ffe26b01a8107ee654

**Comment by dvarrazzo** [comment]
Cool!

Are you able to provide a MR to use it in psycopg?

**Comment by JelteF** [comment]
Can this issue be re-opened?

**Comment by dvarrazzo** [comment]
Hi @JelteF! Sorry, I missed the action when you wrote back in July.

**Comment by dvarrazzo** [comment]
Out of curiosity, has there been any progress in the issue of detecting pgbouncer on connection?

If we could sniff that we could take proper precautions, such as making sure that libpq is >=17 in order to use the Close message, or possibly other customisations.

**Comment by JelteF** [comment]
Not yet... Had too many other things to do so far

**Comment by adriangb** [comment]
Hi folks, I just ran into this and it's unfortunate because it seems like I have to pick between prepared statements and transaction pooling. Have there been any updates to get this implemented, or do either of you have any suggestions for workarounds?

One thought I had is that it may not be necessary to detect that users are connected to pgbouncer (or really any othe connection pooler, e.g. pgcat probably operates the same): users probably know when they are and can set a connection kwarg or similar.

**Comment by adriangb** [comment]
One option would be to add a config knob to just not call DEALLOCATE: https://www.crunchydata.com/blog/prepared-statements-in-transaction-mode-for-pgbouncer#caveat-two-beware-of-deallocation

**Comment by JelteF** [comment]
> One option could be to add a config knob to just not call DEALLOCATE: https://www.crunchydata.com/blog/prepared-statements-in-transaction-mode-for-pgbouncer#caveat-two-beware-of-deallocation

You can do this by setting prepared_max to None or float('infinity') I think:

https://www.psycopg.org/psycopg3/docs/api/connections.html#psycopg.Connection.prepared_max

**Comment by dvarrazzo** [comment]
@JelteF FYI I'm taking a look to this feature request again. So if it's on the radar for you too please let's organise the work.

The first round will try to use the My first idea would be to use the PQclosePrepared function instead of DEALLOCATE.

Second round might be to look into sending a bunch of Close, maybe in a pipeline, instead of DEALLOCATE ALL, and see if it's not too bad. Or has there been some progress in pgBouncer with parsing the DEALLOCATE ALL statement?

Thank you very much

**Comment by JelteF** [comment]
It wasn't really on my radar at this point. But DEALLOCATE ALL is now supported by PgBouncer (plain DEALLOCATE still not)

**Comment by dvarrazzo** [comment]
> It wasn't really on my radar at this point. But DEALLOCATE ALL is now supported by PgBouncer (plain DEALLOCATE still not)

Thank you @JelteF, this is a good piece of information. Could you also please clarify:

- from which PgBouncer release is protocol-level deallocation (via `PQsendClosePrepared`) supported?
- from which PgBouncer release is DEALLOCATE ALL supported?

Thank you very much!

**Comment by JelteF** [comment]
>     * from which PgBouncer release is protocol-level deallocation (via `PQsendClosePrepared`) supported?

1.21.0

>     * from which PgBouncer release is DEALLOCATE ALL supported?

1.22.0

**Comment by dvarrazzo** [comment]
Ok, so reimplementing DEALLOCATE ALL with a loop of deallocate seems would only be useful for the window 1.21-1.22. I think we can avoid to do that, and declare something like "prepared statements with PgBouncer are only supported with libpq >= 17 and PgBouncer 1.22".

Does it sound ok?

**Comment by adriangb** [comment]
> > One option could be to add a config knob to just not call DEALLOCATE: https://www.crunchydata.com/blog/prepared-statements-in-transaction-mode-for-pgbouncer#caveat-two-beware-of-deallocation
> 
> You can do this by setting prepared_max to None or float('infinity') I think:
> 
> https://www.psycopg.org/psycopg3/docs/api/connections.html#psycopg.Connection.prepared_max

That disables preprared statements. I'd like to be able to use them :)

**Comment by adriangb** [comment]
> Does it sound ok?

That sounds reasonable to me. One thing that's not clear to me: libpq is developed in lockstep with postgres, but I imagine I can run a psycopg build that uses libpq 17 right now against postgres 16 and it should (generally) work?

**Comment by adriangb** [comment]
By the way, my workaround for now is to monkey patch `Connection._prepare` during `configure()` with this:

```python
class CustomPrepareManager(PrepareManager):
    """Custom PrepareManager that tweaks caching and cleanup.

    1. Removes trailing comments from queries before hashing them so that they can be cached.
    2. Skips deallocate commands in maintenance commands because this would interfere with pgBouncer's prepared statement caching (see https://github.com/psycopg/psycopg/issues/589).
    """

    @staticmethod
    def key(query: PostgresQuery) -> tuple[bytes, tuple[int, ...]]:
        # remove trailing comments from the query (SQLCommenter style trace info)
        decoded_query = query.query.decode()
        query_str = _TRAILING_COMMENT_REMOVER.sub('', decoded_query).strip().encode()
        if decoded_query[-1] == ';':
            query_str += b';'
        return (query_str, query.types)

    def get_maintenance_commands(self) -> Iterator[bytes]:
        return (
            command for command in super().get_maintenance_commands() if 'DEALLOCATE' not in command.decode().upper()
        )
```

(the comments thing is unrelated, it'd be nice if both psycopg and pgbouncer could ignore comments but I don't see how that's a reasonable feature other than providing a user hook / regex for users to write a targeted version like this one)

**Comment by JelteF** [comment]
> That disables preprared statements. I'd like to be able to use them :)

Setting `prepared_max` to `float('infinity')` enables prepared statements, but effectively disables the `DEALLOCTE` logic.

**Comment by JelteF** [comment]
> but I imagine I can run a psycopg build that uses libpq 17 right now against postgres 16 and it should (generally) work?

Yes that should work fine

**Comment by JelteF** [comment]
> Ok, so reimplementing DEALLOCATE ALL with a loop of deallocate seems would only be useful for the window 1.21-1.22. I think we can avoid to do that, and declare something like "prepared statements with PgBouncer are only supported with libpq >= 17 and PgBouncer 1.22".
> 
> Does it sound ok?

Sounds good, maybe mention about the `float('infinity')` trick as well, since PG17 is not released yet, so people are unlikely to use libpq17 at the moment.

**Comment by adriangb** [comment]
> Setting `prepared_max` to `float('infinity')` enables prepared statements, but effectively disables the `DEALLOCTE` logic.

Ah, nice! Much better than my hack above. I assume then that I do have to make sure that the number of prepared statements doesn't grow unbounded (like it would if I didn't have my remove trailing comment hack above) in my application code or pgbouncer, and rely on `max_lifetime ` in psycopg and `server_lifetime` in pgbouncer to do gc?

**Comment by dvarrazzo** [comment]
> One thing that's not clear to me: libpq is developed in lockstep with postgres, but I imagine I can run a psycopg build that uses libpq 17 right now against postgres 16 and it should (generally) work?

Yes, that's correct. The only thing that libpq adds is an interface to send a Close command on its own, but the underlying client-server communication protocol had that command implemented for a long time.

What will happen is that we would bundle libpq 17 as soon as it's available, therefore the feature would be usable installing just `psycopg[binary]`, with the right PgBouncer at the receiving end.

I assume there is no way to detect PgBouncer version to make our consideration about the features we can use.

**Comment by dvarrazzo** [comment]
@adriangb 

```python
    def get_maintenance_commands(self) -> Iterator[bytes]:
        return (
            command for command in super().get_maintenance_commands() if 'DEALLOCATE' not in command.decode().upper()
        )
```

you may as well make this method always return an empty list: it will only contain DEALLOCATE statements. However this will break with sequences of commands involving dropping objects (explicitly or with a rollback), as per conversation linked in the top message.

Needless to say, if you do something like that, pin your library version well, because you are using internal undocumented interface and we will accept no semver police if they break (matter of fact, I am right now dropping `get_maintenance_commands()`).

**Comment by adriangb** [comment]
Maybe I'll just wait for next release with the fix 😅, I just thought it was interesting that patch at least "works" at first glance.

**Comment by dvarrazzo** [comment]
> Sounds good, maybe mention about the `float('infinity')` trick as well, since PG17 is not released yet, so people are unlikely to use libpq17 at the moment.

Well, this actually works even with the current Psycopg implementation and PgBouncer 1.22. So it's a hint that we could already add to the documentation.

My preference is for `sys.maxsize` though :wink:

**Comment by dvarrazzo** [comment]
A MR to add libpq support is available in #773. Review is welcome :slightly_smiling_face:
